src/test/kotlin/tools/TwoDimensionalPoint.kt

```diff
@@ -0,0 +1,183 @@
+package tools
+
+import kotlin.math.abs
+
+data class Point(val x: Int, val y: Int, var render: Char = 'x') {
+
+  fun manhattanDistance(other: Point): Int {
+    return abs(this.x - other.x) + abs(this.y - other.y)
+  }
+
+  fun directNeighbors(): List<Point> {
+    return Direction.values()
+        .filter {
+          it != Direction.DownRight &&
+              it != Direction.UpRight &&
+              it != Direction.UpLeft &&
+              it != Direction.DownLeft
+        }
+        .map { this.move(it) }
+        .filter { it.x >= 0 && it.y >= 0 }
+        .toList()
+  }
+
+  fun move(direction: Direction): Point {
+    return when (direction) {
+      Direction.Right -> Point(this.x + 1, this.y, render)
+      Direction.Up -> Point(this.x, this.y + 1, render)
+      Direction.Left -> Point(this.x - 1, this.y, render)
+      Direction.Down -> Point(this.x, this.y - 1, render)
+      Direction.DownLeft -> Point(x - 1, y - 1, render)
+      Direction.DownRight -> Point(x + 1, y - 1, render)
+      Direction.UpLeft -> Point(x - 1, y + 1, render)
+      Direction.UpRight -> Point(x + 1, y + 1, render)
+    }
+  }
+
+  fun render(): Char {
+    return render
+  }
+
+  override fun equals(other: Any?): Boolean {
```

Comment:
data classes already have both `hashCode` and `equals`

Reply:
Yes, but I wanted to exclude the rendered char from equals. A point which is rendered as `x` should be equal to a point at the same coords rendered as `#`.